crates/staking/README.md

```diff
@@ -0,0 +1,139 @@
+# Staking Pallet
+
+This pallet is used to manage joint funds for Vaults and Nominators.
+
+## Slashing
+
+We extend the [scalable reward distribution](https://solmaz.io/2019/02/24/scalable-reward-changing/) to account for the imbalance in Vault collateral. Instead of tallying the `reward_per_token` we increase `slash_per_token` to lazily calculate the proportion of collateral to subtract from a Vault or Nominator's stake on deposit or withdrawal.
+
+We present the simple solution (without rewards) below:
+
+```python
+class Slashing:
+    def __init__(self):
+        self.total_stake = 0
+        self.slash_per_token = 0
+        self.stake = {}
+        self.slash_tally = {}
+
+    def deposit_stake(self, address, amount):
+        if address not in self.stake:
+            self.stake[address] = 0
+            self.slash_tally[address] = 0
+
+        self.stake[address] += amount
+        self.slash_tally[address] += self.slash_per_token * amount
+        self.total_stake += amount
+
+    def slash_stake(self, amount):
+        self.slash_per_token += amount / self.total_stake
+
+    def compute_stake(self, address):
+        to_slash = self.stake[address] * self.slash_per_token - self.slash_tally[address]
+        return self.stake[address] - to_slash
+
+    def withdraw_stake(self, address, amount):
+        if amount > self.compute_stake(address):
+            raise Exception("Invalid amount")
+
+        self.stake[address] -= amount
+        self.slash_tally[address] -= self.slash_per_token * amount
+        self.total_stake -= - amount
+```
+
+Without accounting for changing stake sizes we can reduce the staking calculation to the following formula:
+
+```
+actual_stake = stake - (stake * (slashed_amount / total_stake))
+```
+
+### Example
+
+1. Alice deposits 100 DOT in staking pool A.
+2. Bob deposits 50 DOT in staking pool A.
+
+At this point in time, assuming we slash 50 DOT in A, we have the following stakes:
+
+- Alice may withdraw `100 - (100 * (50 / 150)) = 66.66 DOT`.
+- Bob may withdraw `50 - (50 * (50 / 150)) = 33.33 DOT`.
+
+### Unslashing
+
+It is reasonable to assume that stake may be re-distributed to participants after slashing. In this case we would want to "unslash" participant stake proportionally. Since we calculate the amount to slash dynamically we can handle this case by reducing `slash_per_token` by the amount to release.
+
+## Rewards
```

Comment:
I think it's better to understand `slashing` first since that would determine the rewards earned. Is there a particular reason why you think rewards should go first?

Reply:
For me the slashing thing is an extension to the reward calculation scheme. But sure, I see your point